content/en/blog/_posts/2024-04-02-introducing-jobset/2024-04-02-introducing-jobset.md

```diff
@@ -0,0 +1,175 @@
+
+```yaml
+# Distributed training of a traditional CNN model to do image classification 
+# using the MNIST dataset and PyTorch.
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: pytorch
+spec:
+  replicatedJobs:
+  - name: workers
+    template:
+      spec:
+        parallelism: 4
+        completions: 4
+        backoffLimit: 0
+        template:
+          spec:
+            containers:
+            - name: pytorch
+              image: gcr.io/k8s-staging-jobset/pytorch-mnist:latest
+              ports:
+              - containerPort: 3389
+              env:
+              - name: MASTER_ADDR
+                value: "pytorch-workers-0-0.pytorch"
+              - name: MASTER_PORT
+                value: "3389"
+              - name: RANK
+                valueFrom:
+                  fieldRef:
+                    fieldPath: metadata.annotations['batch.kubernetes.io/job-completion-index']
+              # Force python to not buffer output and write directly to stdout, so we can view training logs via `kubectl logs`.
+              - name: PYTHONUNBUFFERED
+                value: "0"
+              command:
+              - bash
+              - -xc
+              - |
+                torchrun --rdzv_id=123 --nnodes=4 --nproc_per_node=1 --master_addr=$MASTER_ADDR --master_port=$MASTER_PORT --node_rank=$RANK mnist.py --epochs=1 --log-interval=1 
+```
+
+Leader-Worker
+
+A popular paradigm for distributed workloads is the leader-worker design. Leader-worker entails having a leader pod, with a different template than workers, and worker pods. The leader must be started before the workers are started and if all the workers finish we can consider the workload complete and the overallworkload can be considered successful.
+
+a) Different pod templates for workers and leaders
+b) Success policies on the workers
+c) Startup policies
+
+
+Example workload
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: success-policy
+spec:
+# We want to start our JobSet in order with leader starting first and then workers starting
+  startupPolicy:
+    startupPolicyOrder: InOrder
+# We want to declare our JobSet successful if workers finish.
+# If workers finish we should clean up the remaining replicatedJobs.
+  successPolicy:
+    operator: All
+    targetReplicatedJobs:
+    - workers
+  replicatedJobs:
+  - name: leader
+    replicas: 1
+    template:
+      spec:
+        # Set backoff limit to 0 so job will immediately fail if any pod fails.
+        backoffLimit: 0 
+        completions: 1
+        parallelism: 1
+        template:
+          spec:
+            containers:
+            - name: leader
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10000
+  - name: workers
```

Comment:
are there any tangible examples of this? I assume not if you're using bash here, but I find it odd this is abstract

Reply:
I agree, we should make the example more concrete.

Reply:
I think we can probably come up with a GPU based example and simply use fake node labels like "cloud.provider.com/gpu-label" instead of a vendor specific one

Reply:
Maybe you can link to GKE related use cases as concrete examples? 

If @haircommander point is taking literal, we should show concrete usecases also.

Reply:
What do you think about that @sftim? Should we try to keep the post self contained?